### DISCUSSION TITLE
Selecting all columns of a particular type?

### DISCUSSION THREAD

**Discussion by bzm3r** [duckdb/duckdb#13675] [Q&A]
Suppose I want to select all columns of a certain type from a DuckDB table. For example, selecting all `VARCHAR` type columns, after creating a table like:

```sql
CREATE TABLE dummy (x VARCHAR, y BIGINT, z VARCHAR);
INSERT INTO dummy
VALUES ('a', 0, 'a'),
  ('b', 1, 'b'),
  ('c', 2, 'c');
```

In general, I might have an arbitrary number of `VARCHAR` type columns, so this query should be "dynamic". I get a list of the relevant columns using `DESCRIBE`:

```
SELECT column_name
FROM (DESCRIBE dummy)
WHERE column_type = 'VARCHAR';
```

This statement gives me a list of the column names which have type `VARCHAR`. But how do I use this? I tried using the `COLUMNS` expression:

```
SELECT COLUMNS(
    c->c IN (
      SELECT column_name
      FROM (DESCRIBE dummy)
      WHERE column_type = 'VARCHAR'
    )
  )
FROM dummy
```

But this gives me the error: `BinderException: Binder Error: Table function cannot contain subqueries`. I don't really understand the error. I get the same error when trying:

```sql
SELECT COLUMNS(
    c->list_contains(
      (
        SELECT column_name
        FROM (DESCRIBE dummy)
        WHERE column_type = 'VARCHAR'
      ),
      c
    )
  )
FROM dummy
```

How do I connect the dots between getting a list of columns by querying `DESCRIBE tbl`, and then using that list to select from `tbl`?

**Comment by soerenwolfers** [comment]
previously discussed at https://github.com/duckdb/duckdb/discussions/6706

unfortunately, no conclusion, except the python bindings now have a `select_types` function in case you use those.

**Comment by cmdlineluser** [comment]
Variables were added in 1.1.0

```python
duckdb.sql("""
set variable column_names = (
   from (describe from dummy)
   select list(column_name)
   where column_type.lower() = 'varchar'
)
""")
```
```python
duckdb.sql("""
from dummy select columns(x -> x in getvariable('column_names'))
""")
```
```
┌─────────┬─────────┐
│    x    │    z    │
│ varchar │ varchar │
├─────────┼─────────┤
│ a       │ a       │
│ b       │ b       │
│ c       │ c       │
└─────────┴─────────┘
```

But I don't think there's a way to run `set variable` in a macro or `query()` ...?
